### DISCUSSION TITLE
EC data protection question

### DISCUSSION THREAD

**Discussion by struthio** [moosefs/moosefs#606] [Q&A]
hi,

This is probably silly question, but I would like to fully understand how MooseFS handle EC recovery.
In release notes of Version 4 it is stated:
"EC4+1: Uses 125% of space
while maintaining the same level of hardware failure protection as standard replication (200%)."

For EC4+1 we got 5 'parts', 4 data parts and 1 checksum/parity. This 'same level of protection' is understood as if '1 of 5 chunks' is gone we can recreate original data, but when 2 chunks are gone - data is not possible to be restored ?

Example, consider 2 chunk-servers A and B. We set Storage Class: @4+1,A,B:s. When chunkserver B is gone, everything works as data is still available in chunksever A, just new parity have to be calculated. When chunkserver A is gone, the data is gone as there is too much information lost to restore original data ?

In example above it is till preferred to keep A,B to keep safety of files ?
Are there any 'good practices' on how many chunkservers we should have to keep data relatively save? (like for @4+1 at least 5 chunk servers are recommended) ?

Also how is 'data chunks' managed internally by MooseFS ?
Is there any special processing to keep them in different chunkservers  (and if not possible at least different disks on same chunkserver) ?

**Comment by chogata** [comment]
2 copies - one copy can go missing and the system will be able to restore it from the other copy; 2 copies go missing, the whole chunk is missing
EC4+1 - 5 parts, 4 data parts and 1 parity part. One part (any of them) can go missing and the system will be able to restore the missing part using the other 4. Two parts go missing (3 are left) - nothing can be restored, the whole chunk goes missing
So the security is exactly the same with these two policies, one chunk server can go missing.
For EC format it doesn't matter which part will go missing, the data part or the parity part. As long as there are 4 parts of any kind for EC4 (or 8 parts of any kind for EC8), the system is always ready to restore the missing part. The only difference is, that if it's the data part that goes missing and there is I/O request for the chunk, the I/O will be delayed a bit, because the system needs to restore this part first. If it's the parity part that goes missing, the I/O can happen immediately and the parity part will be restored after I/O is done.

This is all independent of labels, you can just keep the parts anywhere or you can keep data parts on label A and parity parts on label B. Just be careful when you don't have spare servers (aka you have only the minimum required for EC) and you use the "strict" policy, because then one server failure will mean reverting to copies most of the time.

The minimum number of chunk servers for EC4+1 is 6, and for EC8+1 is 10. We recommend to have more (+2 gives the system much more flexibility). EC parts are always kept on different servers, the only exception is an emergency situation, when too many chunk servers go down and a 2 parts need to be recorded on the same server before convertion to copy format is done. You don't have to specify it in any way, it's the same as with copies: 2 copies on 2 different chunk servers, 5 parts on 5 different chunk servers. Otherwise there would be no redundancy ;)
